src/base64.cr

```diff
@@ -110,36 +113,49 @@ module Base64
   end
 
   private def from_base64(data)
-    buf = 0
-    mod = 0
+    len = data.length
     dt = DECODE_TABLE.buffer
-    data.each do |byte|
-      dec = dt[byte]
-      if dec < 0
-        next if dec == -2
-        break if dec == -3
-        raise Error.new("Invalid character '#{byte.chr}'")
-      end
-      buf = (buf | dec) << 6
-      mod += 1
-      if mod == 4
-        mod = 0
-        yield (buf >> 22).to_u8
-        yield (buf >> 14).to_u8
-        yield (buf >> 6).to_u8
+    cstr = data.pointer(len)
+    while (sym = cstr[len - 1]) && (sym == NL || sym == NR || sym == PAD) && (len > 0)
+      len -= 1
+    end
+    endcstr = cstr + len - 4
+
+    while cstr <= endcstr
+      a, b, c, d = next_decoded_value, next_decoded_value, next_decoded_value, next_decoded_value
+
+      yield (a << 2 | b >> 4).to_u8
+      yield (b << 4 | c >> 2).to_u8
+      yield (c << 6 | d).to_u8
+
+      while (cstr.value == NL || cstr.value == NR) && cstr <= endcstr
+        cstr += 1
       end
     end
 
+    mod = (endcstr - cstr) % 4
     if mod == 2
-      yield (buf >> 10).to_u8
+      a, b = next_decoded_value, next_decoded_value
+      yield (a << 2 | b >> 4).to_u8
     elsif mod == 3
-      yield (buf >> 16).to_u8
-      yield (buf >> 8).to_u8
+      a, b, c = next_decoded_value, next_decoded_value, next_decoded_value
+      yield (a << 2 | b >> 4).to_u8
+      yield (b << 4 | c >> 2).to_u8
     elsif mod != 0
       raise Error.new("Wrong length")
     end
   end
 
+  private macro next_decoded_value
```

Comment:
why not just inline closure without copying, if closure used only in this method.